File: p4.pl
#!/usr/bin/perl
use warnings;
use strict;
use Time::HiRes qw(gettimeofday tv_interval);

# Euler Project #4
# Find the largest palindrome made from the product of two 3-digit numbers.

my $t0 = [gettimeofday()];
my $biggest = 0;

for my $i (100..999) {
	for my $j (100..999) {
		my $product = $i * $j;
		my @prod = split //, $product;
		my @rev = reverse @prod;
		#print "@prod  @rev";
		if ($prod[0] == $rev[0]) {
			if ($prod[1] == $rev[1]) {
				$biggest = $product if ($prod[2] == $rev[2] && $product > $biggest);
			}
		}
	}
}
my $elapsed = tv_interval($t0, [gettimeofday()]);

print "$biggest\n";
print "Finished in $elapsed seconds."

File: c++/p4.cpp
// Find the largest palindrom made from the product of two 3-digit numbers.
#include <iostream>
#include <string>
#include <sstream>

std::string toString(int& i) {
	std::stringstream out;
	out << i;
	return (out.str());
}

int main()
{
	using namespace std;

	int biggest = 0;

	for (int i=100; i < 1000; ++i)
	{
		for (int j=100; j < 1000; ++j)
		{
			int product = i * j;
			if (product < biggest)
				continue;
			
			string forwards = toString(product);
			string backwards = string(forwards.rbegin(), forwards.rend());

			if (forwards[0] == backwards[0] && forwards[1] == backwards[1])
			{
				if (forwards[2] == backwards[2] && product > biggest)
					biggest = product;
			}
		}
	}

	cout << biggest << endl;
}
